File: app/src/main/java/io/github/anonymous/pepper_realtime/ui/settings/SettingsViewModel.java
package io.github.anonymous.pepper_realtime.ui.settings;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.Set;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.github.anonymous.pepper_realtime.manager.SettingsRepository;
import io.github.anonymous.pepper_realtime.network.RealtimeApiProvider;

@HiltViewModel
public class SettingsViewModel extends ViewModel {

    private final SettingsRepository settingsRepository;

    // Events
    private final MutableLiveData<Boolean> restartSessionEvent = new MutableLiveData<>();
    private final MutableLiveData<Boolean> updateSessionEvent = new MutableLiveData<>();
    private final MutableLiveData<Boolean> restartRecognizerEvent = new MutableLiveData<>();
    private final MutableLiveData<Integer> volumeChangeEvent = new MutableLiveData<>();

    @Inject
    public SettingsViewModel(SettingsRepository settingsRepository) {
        this.settingsRepository = settingsRepository;
    }

    // Event Getters
    public LiveData<Boolean> getRestartSessionEvent() {
        return restartSessionEvent;
    }

    public LiveData<Boolean> getUpdateSessionEvent() {
        return updateSessionEvent;
    }

    public LiveData<Boolean> getRestartRecognizerEvent() {
        return restartRecognizerEvent;
    }

    public LiveData<Integer> getVolumeChangeEvent() {
        return volumeChangeEvent;
    }

    // Settings Operations

    public void setSystemPrompt(String prompt) {
        if (!prompt.equals(settingsRepository.getSystemPrompt())) {
            settingsRepository.setSystemPrompt(prompt);
            updateSessionEvent.setValue(true);
        }
    }

    public void setModel(String model) {
        if (!model.equals(settingsRepository.getModel())) {
            settingsRepository.setModel(model);
            restartSessionEvent.setValue(true);
        }
    }

    public void setVoice(String voice) {
        if (!voice.equals(settingsRepository.getVoice())) {
            settingsRepository.setVoice(voice);
            restartSessionEvent.setValue(true);
        }
    }

    public void setSpeedProgress(int progress) {
        if (progress != settingsRepository.getSpeedProgress()) {
            settingsRepository.setSpeedProgress(progress);
            updateSessionEvent.setValue(true);
        }
    }

    public void setLanguage(String language) {
        if (!language.equals(settingsRepository.getLanguage())) {
            settingsRepository.setLanguage(language);
            restartRecognizerEvent.setValue(true);
        }
    }

    public void setApiProvider(String provider) {
        if (!provider.equals(settingsRepository.getApiProvider())) {
            settingsRepository.setApiProvider(provider);
            restartSessionEvent.setValue(true);
        }
    }

    public void setAudioInputMode(String mode) {
        if (!mode.equals(settingsRepository.getAudioInputMode())) {
            settingsRepository.setAudioInputMode(mode);
            restartSessionEvent.setValue(true);
        }
    }

    public void setTemperatureProgress(int progress) {
        if (progress != settingsRepository.getTemperatureProgress()) {
            settingsRepository.setTemperatureProgress(progress);
            updateSessionEvent.setValue(true);
        }
    }

    public void setVolume(int volume) {
        if (volume != settingsRepository.getVolume()) {
            settingsRepository.setVolume(volume);
            volumeChangeEvent.setValue(volume);
        }
    }

    public void setSilenceTimeout(int timeout) {
        if (timeout != settingsRepository.getSilenceTimeout()) {
            settingsRepository.setSilenceTimeout(timeout);
            restartRecognizerEvent.setValue(true);
        }
    }

    public void setConfidenceThreshold(float threshold) {
        if (threshold != settingsRepository.getConfidenceThreshold()) {
            settingsRepository.setConfidenceThreshold(threshold);
            // No immediate action needed, used by recognizer on next result
        }
    }

    public void setEnabledTools(Set<String> tools) {
        if (!tools.equals(settingsRepository.getEnabledTools())) {
            settingsRepository.setEnabledTools(tools);
            updateSessionEvent.setValue(true);
        }
    }

    // Realtime API Specific Settings

    public void setTranscriptionModel(String model) {
        if (!model.equals(settingsRepository.getTranscriptionModel())) {
            settingsRepository.setTranscriptionModel(model);
            triggerRealtimeSettingChange();
        }
    }

    public void setTranscriptionLanguage(String language) {
        if (!language.equals(settingsRepository.getTranscriptionLanguage())) {
            settingsRepository.setTranscriptionLanguage(language);
            triggerRealtimeSettingChange();
        }
    }

    public void setTranscriptionPrompt(String prompt) {
        if (!prompt.equals(settingsRepository.getTranscriptionPrompt())) {
            settingsRepository.setTranscriptionPrompt(prompt);
            triggerRealtimeSettingChange();
        }
    }

    public void setTurnDetectionType(String type) {
        if (!type.equals(settingsRepository.getTurnDetectionType())) {
            settingsRepository.setTurnDetectionType(type);
            triggerRealtimeSettingChange();
        }
    }

    public void setVadThreshold(float threshold) {
        if (threshold != settingsRepository.getVadThreshold()) {
            settingsRepository.setVadThreshold(threshold);
            triggerRealtimeSettingChange();
        }
    }

    public void setPrefixPadding(int padding) {
        if (padding != settingsRepository.getPrefixPadding()) {
            settingsRepository.setPrefixPadding(padding);
            triggerRealtimeSettingChange();
        }
    }

    public void setSilenceDuration(int duration) {
        if (duration != settingsRepository.getSilenceDuration()) {
            settingsRepository.setSilenceDuration(duration);
            triggerRealtimeSettingChange();
        }
    }

    public void setIdleTimeout(int timeout) {
        Integer current = settingsRepository.getIdleTimeout();
        if (current == null || timeout != current) {
            settingsRepository.setIdleTimeout(timeout);
            triggerRealtimeSettingChange();
        }
    }

    public void setEagerness(String eagerness) {
        if (!eagerness.equals(settingsRepository.getEagerness())) {
            settingsRepository.setEagerness(eagerness);
            triggerRealtimeSettingChange();
        }
    }

    public void setNoiseReduction(String reduction) {
        if (!reduction.equals(settingsRepository.getNoiseReduction())) {
            settingsRepository.setNoiseReduction(reduction);
            triggerRealtimeSettingChange();
        }
    }

    private void triggerRealtimeSettingChange() {
        if (settingsRepository.isUsingRealtimeAudioInput()) {
            updateSessionEvent.setValue(true);
        }
    }

    // Getters for UI Initialization
    public String getSystemPrompt() {
        return settingsRepository.getSystemPrompt();
    }

    public String getModel() {
        return settingsRepository.getModel();
    }

    public String getVoice() {
        return settingsRepository.getVoice();
    }

    public int getSpeedProgress() {
        return settingsRepository.getSpeedProgress();
    }

    public String getLanguage() {
        return settingsRepository.getLanguage();
    }

    public String getApiProvider() {
        return settingsRepository.getApiProvider();
    }

    public String getAudioInputMode() {
        return settingsRepository.getAudioInputMode();
    }

    public int getTemperatureProgress() {
        return settingsRepository.getTemperatureProgress();
    }

    public int getVolume() {
        return settingsRepository.getVolume();
    }

    public int getSilenceTimeout() {
        return settingsRepository.getSilenceTimeout();
    }

    public float getConfidenceThreshold() {
        return settingsRepository.getConfidenceThreshold();
    }

    public Set<String> getEnabledTools() {
        return settingsRepository.getEnabledTools();
    }

    public String getTranscriptionModel() {
        return settingsRepository.getTranscriptionModel();
    }

    public String getTranscriptionLanguage() {
        return settingsRepository.getTranscriptionLanguage();
    }

    public String getTranscriptionPrompt() {
        return settingsRepository.getTranscriptionPrompt();
    }

    public String getTurnDetectionType() {
        return settingsRepository.getTurnDetectionType();
    }

    public float getVadThreshold() {
        return settingsRepository.getVadThreshold();
    }

    public int getPrefixPadding() {
        return settingsRepository.getPrefixPadding();
    }

    public int getSilenceDuration() {
        return settingsRepository.getSilenceDuration();
    }

    public Integer getIdleTimeout() {
        return settingsRepository.getIdleTimeout();
    }

    public String getEagerness() {
        return settingsRepository.getEagerness();
    }

    public String getNoiseReduction() {
        return settingsRepository.getNoiseReduction();
    }

    // Event consumption methods for ChatActivity
    public void consumeRestartSessionEvent() {
        restartSessionEvent.setValue(false);
    }

    public void consumeUpdateSessionEvent() {
        updateSessionEvent.setValue(false);
    }

    public void consumeRestartRecognizerEvent() {
        restartRecognizerEvent.setValue(false);
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/ui/settings/SettingsViewModel.kt
package io.github.anonymous.pepper_realtime.ui.settings

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.anonymous.pepper_realtime.manager.SettingsRepository
import javax.inject.Inject

@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val settingsRepository: SettingsRepository
) : ViewModel() {

    // Events
    private val _restartSessionEvent = MutableLiveData<Boolean>()
    private val _updateSessionEvent = MutableLiveData<Boolean>()
    private val _restartRecognizerEvent = MutableLiveData<Boolean>()
    private val _volumeChangeEvent = MutableLiveData<Int>()

    // Event Getters
    val restartSessionEvent: LiveData<Boolean> get() = _restartSessionEvent
    val updateSessionEvent: LiveData<Boolean> get() = _updateSessionEvent
    val restartRecognizerEvent: LiveData<Boolean> get() = _restartRecognizerEvent
    val volumeChangeEvent: LiveData<Int> get() = _volumeChangeEvent

    // Batch Update Mode
    private var isBatchMode = false
    private var pendingRestart = false
    private var pendingUpdate = false
    private var pendingRecognizerRestart = false
    
    fun beginEditing() {
        isBatchMode = true
        pendingRestart = false
        pendingUpdate = false
        pendingRecognizerRestart = false
    }
    
    fun commitChanges() {
        isBatchMode = false
        if (pendingRestart) _restartSessionEvent.value = true
        else if (pendingUpdate) _updateSessionEvent.value = true
        
        if (pendingRecognizerRestart) _restartRecognizerEvent.value = true
        
        // Reset pending flags
        pendingRestart = false
        pendingUpdate = false
        pendingRecognizerRestart = false
    }
    
    // Settings Operations
    fun setSystemPrompt(prompt: String) {
        if (prompt != settingsRepository.systemPrompt) {
            settingsRepository.systemPrompt = prompt
            if (isBatchMode) pendingUpdate = true else _updateSessionEvent.value = true
        }
    }

    fun setModel(model: String) {
        if (model != settingsRepository.model) {
            settingsRepository.model = model
            if (isBatchMode) pendingRestart = true else _restartSessionEvent.value = true
        }
    }

    fun setVoice(voice: String) {
        if (voice != settingsRepository.voice) {
            settingsRepository.voice = voice
            if (isBatchMode) pendingRestart = true else _restartSessionEvent.value = true
        }
    }

    fun setSpeedProgress(progress: Int) {
        if (progress != settingsRepository.speedProgress) {
            settingsRepository.speedProgress = progress
            if (isBatchMode) pendingUpdate = true else _updateSessionEvent.value = true
        }
    }

    fun setLanguage(language: String) {
        if (language != settingsRepository.language) {
            settingsRepository.language = language
            if (isBatchMode) pendingRecognizerRestart = true else _restartRecognizerEvent.value = true
        }
    }

    fun setApiProvider(provider: String) {
        if (provider != settingsRepository.apiProvider) {
            settingsRepository.apiProvider = provider
            if (isBatchMode) pendingRestart = true else _restartSessionEvent.value = true
        }
    }

    fun setAudioInputMode(mode: String) {
        if (mode != settingsRepository.audioInputMode) {
            settingsRepository.audioInputMode = mode
            if (isBatchMode) pendingRestart = true else _restartSessionEvent.value = true
        }
    }

    fun setTemperatureProgress(progress: Int) {
        if (progress != settingsRepository.temperatureProgress) {
            settingsRepository.temperatureProgress = progress
            if (isBatchMode) pendingUpdate = true else _updateSessionEvent.value = true
        }
    }

    fun setVolume(volume: Int) {
        if (volume != settingsRepository.volume) {
            settingsRepository.volume = volume
            _volumeChangeEvent.value = volume // Volume should change immediately for feedback
        }
    }

    fun setSilenceTimeout(timeout: Int) {
        if (timeout != settingsRepository.silenceTimeout) {
            settingsRepository.silenceTimeout = timeout
            if (isBatchMode) pendingRecognizerRestart = true else _restartRecognizerEvent.value = true
        }
    }

    fun setConfidenceThreshold(threshold: Float) {
        if (threshold != settingsRepository.confidenceThreshold) {
            settingsRepository.confidenceThreshold = threshold
            // No immediate action needed
        }
    }

    fun setEnabledTools(tools: Set<String>) {
        if (tools != settingsRepository.enabledTools) {
            settingsRepository.enabledTools = tools
            if (isBatchMode) pendingUpdate = true else _updateSessionEvent.value = true
        }
    }
    
    // Realtime API Specific Settings
    fun setTranscriptionModel(model: String) {
        if (model != settingsRepository.transcriptionModel) {
            settingsRepository.transcriptionModel = model
            triggerRealtimeSettingChange()
        }
    }

    fun setTranscriptionLanguage(language: String) {
        if (language != settingsRepository.transcriptionLanguage) {
            settingsRepository.transcriptionLanguage = language
            triggerRealtimeSettingChange()
        }
    }

    fun setTranscriptionPrompt(prompt: String) {
        if (prompt != settingsRepository.transcriptionPrompt) {
            settingsRepository.transcriptionPrompt = prompt
            triggerRealtimeSettingChange()
        }
    }

    fun setTurnDetectionType(type: String) {
        if (type != settingsRepository.turnDetectionType) {
            settingsRepository.turnDetectionType = type
            triggerRealtimeSettingChange()
        }
    }

    fun setVadThreshold(threshold: Float) {
        if (threshold != settingsRepository.vadThreshold) {
            settingsRepository.vadThreshold = threshold
            triggerRealtimeSettingChange()
        }
    }

    fun setPrefixPadding(padding: Int) {
        if (padding != settingsRepository.prefixPadding) {
            settingsRepository.prefixPadding = padding
            triggerRealtimeSettingChange()
        }
    }

    fun setSilenceDuration(duration: Int) {
        if (duration != settingsRepository.silenceDuration) {
            settingsRepository.silenceDuration = duration
            triggerRealtimeSettingChange()
        }
    }

    fun setIdleTimeout(timeout: Int) {
        val current = settingsRepository.idleTimeout
        if (current == null || timeout != current) {
            settingsRepository.idleTimeout = timeout
            triggerRealtimeSettingChange()
        }
    }

    fun setEagerness(eagerness: String) {
        if (eagerness != settingsRepository.eagerness) {
            settingsRepository.eagerness = eagerness
            triggerRealtimeSettingChange()
        }
    }

    fun setNoiseReduction(reduction: String) {
        if (reduction != settingsRepository.noiseReduction) {
            settingsRepository.noiseReduction = reduction
            triggerRealtimeSettingChange()
        }
    }
    
    private fun triggerRealtimeSettingChange() {
        if (settingsRepository.isUsingRealtimeAudioInput) {
            if (isBatchMode) pendingUpdate = true else _updateSessionEvent.value = true
        }
    }

    // Getters for UI Initialization
    fun getSystemPrompt(): String = settingsRepository.systemPrompt
    fun getModel(): String = settingsRepository.model
    fun getVoice(): String = settingsRepository.voice
    fun getSpeedProgress(): Int = settingsRepository.speedProgress
    fun getLanguage(): String = settingsRepository.language
    fun getApiProvider(): String = settingsRepository.apiProvider
    fun getAudioInputMode(): String = settingsRepository.audioInputMode
    fun getTemperatureProgress(): Int = settingsRepository.temperatureProgress
    fun getVolume(): Int = settingsRepository.volume
    fun getSilenceTimeout(): Int = settingsRepository.silenceTimeout
    fun getConfidenceThreshold(): Float = settingsRepository.confidenceThreshold
    fun getEnabledTools(): Set<String> = settingsRepository.enabledTools
    fun getTranscriptionModel(): String = settingsRepository.transcriptionModel
    fun getTranscriptionLanguage(): String = settingsRepository.transcriptionLanguage
    fun getTranscriptionPrompt(): String = settingsRepository.transcriptionPrompt
    fun getTurnDetectionType(): String = settingsRepository.turnDetectionType
    fun getVadThreshold(): Float = settingsRepository.vadThreshold
    fun getPrefixPadding(): Int = settingsRepository.prefixPadding
    fun getSilenceDuration(): Int = settingsRepository.silenceDuration
    fun getIdleTimeout(): Int? = settingsRepository.idleTimeout
    fun getEagerness(): String = settingsRepository.eagerness
    fun getNoiseReduction(): String = settingsRepository.noiseReduction

    // Event consumption methods for ChatActivity
    fun consumeRestartSessionEvent() {
        _restartSessionEvent.value = false
    }

    fun consumeUpdateSessionEvent() {
        _updateSessionEvent.value = false
    }

    fun consumeRestartRecognizerEvent() {
        _restartRecognizerEvent.value = false
    }
}

